torch/quantization/quantize.py

```diff
@@ -150,6 +151,11 @@ def prepare_qat(model, qconfig_dict=None):
     model = convert(model, DEFAULT_QAT_MODULE_MAPPING)
     return model
 
+def prepare_dynamic(model, qconfig_dict=None):
+    propagate_qconfig(model, qconfig_dict)
+    add_observer(model)
```

Comment:
Remove add_observer, we do not need it for dynamic quant. We can address the dependency in swap_module by modifying it as follows:
```python
def swap_module(mod, mapping):
    r"""Swaps the module if it has a quantized counterpart and it has an
    `observer` attached.
    Args:
        mod: input module
        mapping: a dictionary that maps from nn module to nnq module
    Return:
        The corresponding quantized module of `mod`
    """
    new_mod = mod
    if hasattr(mod, 'qconfig'): # Instead of observer, use qconfig
        if type(mod) in mapping:
            new_mod = mapping[type(mod)].from_float(mod)

    return new_mod
```

Reply:
no we can't, qconfig can be None. we are using observer to mark whether we want quantization or not right now.

Reply:
As @jerryzh168 mentioned, there will be an issue here.

Reply:
@jerryzh168 and myself discussed this, he is going to land a diff that removes the check for observers in swap module (basically the code below):
So, now you can remove observers as a requirement for dynamic quantization.

def swap_module(mod, mapping):
r"""Swaps the module if it has a quantized counterpart and it has an
    `observer` attached.
    Args:
        mod: input module
        mapping: a dictionary that maps from nn module to nnq module
    Return:
        The corresponding quantized module of `mod`
    """
    new_mod = mod
    if hasattr(mod, 'qconfig') and mod.qconfig is not None: # Instead of observer, use qconfig
        if type(mod) in mapping:
            new_mod = mapping[type(mod)].from_float(mod)

    return new_mod

Reply:
Got it! Will modify this.

Reply:
This has been done by @jerryzh168 in https://github.com/pytorch/pytorch/pull/23561. Updated the PR by removing the `add_observer`.